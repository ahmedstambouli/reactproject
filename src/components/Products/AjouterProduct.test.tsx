// AjouterProduct.test.tsx

import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import AjouterProduct from './AjouterProduct'; // Assurez-vous que le chemin est correct
import axios from 'axios';

// Mocking de l'API Axios
jest.mock('axios');

describe('AjouterProduct Component', () => {
  // Test de la validation des champs vides
  test('Affiche les erreurs de validation lorsque les champs sont vides', async () => {
    render(<AjouterProduct />);

    // Soumettre le formulaire sans remplir les champs
    const submitButton = screen.getByText(/submit/i); // Assurez-vous que le texte est correct
    fireEvent.click(submitButton);

    // Vérification des messages d'erreur
    const productNameError = await screen.findByText(/Le nom du produit est requis/i);
    expect(productNameError).toBeInTheDocument();

    const priceError = await screen.findByText(/Entrez un prix valide/i);
    expect(priceError).toBeInTheDocument();
  });

  // Test de la récupération des catégories via l'API
  test('Récupère et affiche les catégories à partir de l\'API', async () => {
    const mockCategories = [
      { id: '1', name: 'Electronique' },
      { id: '2', name: 'Vêtements' }
    ];

    // Simulation de la réponse de l'API
    axios.get.mockResolvedValueOnce({ data: { $values: mockCategories } });

    render(<AjouterProduct />);

    // Attendre que les catégories soient affichées
    await waitFor(() => {
      expect(screen.getByText(/Electronique/)).toBeInTheDocument();
      expect(screen.getByText(/Vêtements/)).toBeInTheDocument();
    });
  });

  // Test de la mise à jour de l'état du produit avec des entrées valides
  test('Met à jour le nom du produit et le prix', () => {
    render(<AjouterProduct />);

    const productNameInput = screen.getByPlaceholderText(/Entrez le nom du produit/i);
    fireEvent.change(productNameInput, { target: { value: 'Produit Test' } });

    const priceInput = screen.getByPlaceholderText(/Entrez le prix du produit/i);
    fireEvent.change(priceInput, { target: { value: '20' } });

    // Vérification que les valeurs sont bien mises à jour
    expect(productNameInput.value).toBe('Produit Test');
    expect(priceInput.value).toBe('20');
  });

  // Test de l'affichage des erreurs lors de la soumission avec des données invalides
  test('Affiche un message d\'erreur lors de la soumission avec des données invalides', async () => {
    render(<AjouterProduct />);

    const submitButton = screen.getByText(/submit/i);
    fireEvent.click(submitButton);

    // Vérification que les messages d'erreur sont affichés
    await waitFor(() => {
      expect(screen.getByText(/Le nom du produit est requis/i)).toBeInTheDocument();
      expect(screen.getByText(/Entrez un prix valide/i)).toBeInTheDocument();
    });
  });

  // Test de l'intégration avec les snapshots (facultatif)
  test('Snapshot du composant AjouterProduct', () => {
    const { asFragment } = render(<AjouterProduct />);
    expect(asFragment()).toMatchSnapshot();
  });
});
