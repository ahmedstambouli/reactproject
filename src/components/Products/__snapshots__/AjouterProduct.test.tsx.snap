// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AjouterProduct Component Snapshot du composant AjouterProduct 1`] = `
<DocumentFragment>
  <div
    class="rounded-sm border border-stroke bg-white shadow-default dark:border-strokedark dark:bg-boxdark"
  >
    <div
      class="py-6 px-4 md:px-6 xl:px-7.5"
    >
      <h4
        class="text-xl font-semibold text-black dark:text-white"
      >
        Add Product
      </h4>
      <form
        class="flex flex-col gap-5.5 p-6.5"
      >
        <div
          class="grid grid-cols-2 gap-4"
        >
          <div>
            <label
              class="mb-3 block text-black dark:text-white"
            >
              Product Name
            </label>
            <input
              class="w-full rounded-lg border-[1.5px] border-stroke bg-transparent py-3 px-5 text-black outline-none transition focus:border-primary active:border-primary dark:border-form-strokedark dark:bg-form-input dark:text-white dark:focus:border-primary"
              placeholder="Enter product name"
              type="text"
              value=""
            />
          </div>
          <div>
            <label
              class="mb-3 block text-black dark:text-white"
            >
              Price
            </label>
            <input
              class="w-full rounded-lg border-[1.5px] border-stroke bg-transparent py-3 px-5 text-black outline-none transition focus:border-primary active:border-primary dark:border-form-strokedark dark:bg-form-input dark:text-white dark:focus:border-primary"
              placeholder="Enter product price"
              type="number"
              value=""
            />
          </div>
          <div>
            <label
              class="mb-3 block text-black dark:text-white"
            >
              Quantity
            </label>
            <input
              class="w-full rounded-lg border-[1.5px] border-stroke bg-transparent py-3 px-5 text-black outline-none transition focus:border-primary active:border-primary dark:border-form-strokedark dark:bg-form-input dark:text-white dark:focus:border-primary"
              placeholder="Enter quantity"
              type="number"
              value=""
            />
          </div>
          <div>
            <label
              class="mb-3 block text-black dark:text-white"
            >
              Category
            </label>
            <select
              class="relative z-20 w-full appearance-none rounded border border-stroke bg-transparent py-3 px-12 outline-none transition focus:border-primary active:border-primary dark:border-form-strokedark dark:bg-form-input "
              required=""
            >
              <option
                value=""
              >
                Select a category
              </option>
            </select>
          </div>
          <div>
            <label
              class="mb-3 block text-black dark:text-white"
            >
              Description
            </label>
            <textarea
              class="w-full rounded-lg border-[1.5px] border-stroke bg-transparent py-3 px-5 text-black outline-none transition focus:border-primary active:border-primary dark:border-form-strokedark dark:bg-form-input dark:text-white dark:focus:border-primary"
              placeholder="Enter description"
              rows="6"
            />
          </div>
          <div>
            <label
              class="mb-3 block text-black dark:text-white"
            >
              Picture Products
            </label>
            <input
              accept="image/*"
              class="w-full cursor-pointer rounded-lg border-[1.5px] border-stroke bg-transparent outline-none transition file:mr-5 file:border-collapse file:cursor-pointer file:border-0 file:border-r file:border-solid file:border-stroke file:bg-whiter file:py-3 file:px-5 file:hover:bg-primary file:hover:bg-opacity-10 focus:border-primary active:border-primary dark:border-form-strokedark dark:bg-form-input dark:file:border-form-strokedark dark:file:bg-white/30 dark:file:text-white dark:focus:border-primary"
              required=""
              type="file"
            />
          </div>
        </div>
        <input
          class="inline-flex items-center justify-center gap-2.5 rounded-md bg-primary py-4 px-10 text-center font-medium text-white hover:bg-opacity-90 lg:px-8 xl:px-10"
          type="submit"
          value="Add Product"
        />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
